### DISCUSSION TITLE
tquic 如何保存skey文件

### DISCUSSION THREAD

**Discussion by lion117** [Tencent/tquic#302] [Q&A (问题和回答)] [answered]
# 背景
1. ubuntuX64 
2. simple_client.c simaple_server.c 
3. libtquic.a 

# 描述
> 我使用 simple_client.c simaple_server.c 两个demo , 可以正常实现 echo 消息互通.  我希望能够用wireshake 分析网络过程数据. 查阅资料后, 分别在客户端和服务器实现了保存skey的调用. 但是接口一直未正常触发. 想咨询一下我是被卡在哪里了? 

```
void cbSkey(const uint8_t *data,size_t data_len,void *argp)
{
    static  FILE * fd  = NULL;  
    if (!fd)
    {
        fd  = fopen("tquic_keylog_client.log","wb"); 
    }
    fwrite(data, 1,data_len, fd);
    fflush(fd);
}

void client_on_conn_created(void *tctx, struct quic_conn_t *conn) {
    struct simple_client *client = tctx;
    client->conn = conn;

    printf("client_on_conn_created \r\n");

    quic_conn_set_keylog(conn, cbSkey, tctx);
}

```

**Comment by glime-ai** [comment]
根据您提供的代码和描述，您已经在客户端实现了保存密钥日志的回调函数，并在连接创建时设置了该回调函数。然而，接口未正常触发，可能有以下几个原因：

1. **回调函数未被正确调用**：确保在连接创建后，回调函数 `cbSkey` 被正确设置并调用。可以在 `quic_conn_set_keylog` 调用后添加调试信息，确认回调函数是否被设置成功。

2. **连接未成功建立**：确保连接已经成功建立并且有数据传输。可以通过检查连接状态和数据传输情况来确认。

3. **文件权限问题**：确保程序有权限在当前目录下创建和写入 `tquic_keylog_client.log` 文件。可以尝试手动创建该文件并检查权限。

4. **回调函数实现问题**：确保回调函数 `cbSkey` 的实现没有问题。可以在回调函数中添加更多的调试信息，确认函数是否被调用以及数据是否正确写入文件。

以下是一些调试建议：

```c
void cbSkey(const uint8_t *data, size_t data_len, void *argp) {
    static FILE *fd = NULL;
    if (!fd) {
        fd = fopen("tquic_keylog_client.log", "wb");
        if (!fd) {
            perror("Failed to open keylog file");
            return;
        }
    }
    fwrite(data, 1, data_len, fd);
    fflush(fd);
    printf("Keylog data written: %zu bytes\n", data_len); // 添加调试信息
}

void client_on_conn_created(void *tctx, struct quic_conn_t *conn) {
    struct simple_client *client = tctx;
    client->conn = conn;

    printf("client_on_conn_created\n");

    quic_conn_set_keylog(conn, cbSkey, tctx);
    printf("Keylog callback set\n"); // 添加调试信息
}
```

此外，您可以参考以下文档和示例代码，确保您的实现与官方文档一致：

- [QUIC API 参考文档](https://tquic.net/zh/docs/api_reference/c_quic)
- [使用 C 语言接口的教程](https://tquic.net/zh/docs/tutorial/c)
- [GitHub 上的示例代码](https://github.com/tquic-group/tquic-example-c)

希望这些信息能帮助您解决问题。如果问题仍然存在，请提供更多的调试信息以便进一步分析。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by xiaofei0800** [comment, marked as answer]
使用quic_tls_config_new_with_ssl_ctx接口，需要在SSL_CTX中通过SSL_CTX_set_keylog_callback设置相关key log回调。

可以在客户端侧使用quic_tls_config_new_client_config创建tls_config_t结构：
  const char* const protos[1] = {"http/0.9"};
  client->tls_config = quic_tls_config_new_client_config(protos, 1, true);

**Reply by lion117** [reply]
测试有效. 
终于可以分析wireshark抓包解密数据了. 感谢
